app/features/vods/VodRepository.server.ts

```diff
@@ -1,5 +1,258 @@
+import { sql } from "kysely";
 import { db } from "~/db/sql";
+import type { Tables } from "~/db/tables";
+import type {
+	MainWeaponId,
+	ModeShort,
+	StageId,
+} from "~/modules/in-game-lists/types";
+import { weaponIdToArrayWithAlts } from "~/modules/in-game-lists/weapon-ids";
+import {
+	dateToDatabaseTimestamp,
+	dayMonthYearToDatabaseTimestamp,
+} from "~/utils/dates";
+import invariant from "~/utils/invariant";
+import { selectPlayers } from "~/utils/kysely.server";
+import { VODS_PAGE_BATCH_SIZE } from "./vods-constants";
+import type { ListVod, VideoBeingAdded, Vod } from "./vods-types";
+import {
+	extractYoutubeIdFromVideoUrl,
+	hoursMinutesSecondsStringToSeconds,
+} from "./vods-utils";
 
 export function deleteById(id: number) {
 	return db.deleteFrom("UnvalidatedVideo").where("id", "=", id).execute();
 }
+
+export async function findVods({
+	weapon,
+	mode,
+	stageId,
+	type,
+	userId,
+	limit = VODS_PAGE_BATCH_SIZE,
+}: {
+	weapon?: MainWeaponId;
+	mode?: ModeShort;
+	stageId?: StageId;
+	type?: Tables["Video"]["type"];
+	userId?: number;
+	limit?: number;
+}): Promise<Array<ListVod>> {
+	let query = db
+		.selectFrom("Video")
+		.leftJoin("VideoMatch", "Video.id", "VideoMatch.videoId")
+		.leftJoin(
+			"VideoMatchPlayer",
+			"VideoMatch.id",
+			"VideoMatchPlayer.videoMatchId",
+		)
+		.leftJoin("User", "VideoMatchPlayer.playerUserId", "User.id")
+		.selectAll("Video")
+		.select(({ fn, ref }) => [
+			sql<
+				Array<number>
+			>`json_group_array(distinct ${ref("VideoMatchPlayer.weaponSplId")})`
+				.$castTo<MainWeaponId[]>()
+				.as("weapons"),
+			fn
+				.agg("json_group_array", ["VideoMatchPlayer.playerName"])
+				.as("playerNames"),
+			selectPlayers(),
+		]);
+	if (userId) {
+		query = query.where("User.id", "=", userId);
+	} else {
+		if (type) {
+			query = query.where("Video.type", "=", type);
+		}
+		if (mode) {
+			query = query.where("VideoMatch.mode", "=", mode);
+		}
+		if (stageId) {
+			query = query.where("VideoMatch.stageId", "=", stageId);
+		}
+	}
+	if (weapon) {
+		query = query.where((eb) =>
+			eb.or(
+				weaponIdToArrayWithAlts(weapon).map((weapon) => {
+					return eb("VideoMatchPlayer.weaponSplId", "=", weapon);
+				}),
+			),
+		);
+	}
+	const result = query
+		.groupBy("Video.id")
+		.orderBy("Video.youtubeDate", "desc")
+		.limit(limit)
+		.execute();
+
+	const vods = (await result).map((value) => {
+		const { playerNames, players, ...vod } = value;
+		const playerNamesArray: string[] = playerNames as string[];
```

Comment:
We could avoid bypassing the Typescript compiler if we use Kysely more typesafe way. This example is from TournamentRepository:

```ts
jsonArrayFrom(
				eb
					.selectFrom("TournamentStaff")
					.innerJoin("User", "TournamentStaff.userId", "User.id")
					.select([
						...COMMON_USER_FIELDS,
						userChatNameColor,
						"TournamentStaff.role",
					])
					.where("TournamentStaff.tournamentId", "=", id),
			).as("staff"),
```

benefit of writing it like this is that the result is automatically typesafe without any tricks.